kliff/loss.py

```diff
@@ -240,14 +191,11 @@ def __init__(
     ):
 
         default_residual_data = {
-            "energy_weight": 1.0,
-            "forces_weight": 1.0,
-            "stress_weight": 1.0,
             "normalize_by_natoms": True,
         }
 
         residual_data = _check_residual_data(residual_data, default_residual_data)
-        _check_compute_flag(calculator, residual_data)
+        # _check_compute_flag(calculator, residual_data)
```

Comment:
Can you update the `_check_comute_flag` function to something like  `_check_comute_flag(calculator, weight, residual_data)` and check weight there?

Reply:
I commented out `_check_compute_flag` because I am not sure what to check in the residual data. Previously, it checked if the weights are big enough and put a message in the log file if the weights are small. So, when you suggest to update `_check_compute_weight`, is it to keep this functionality (to check if the weight are large enough and give feedback to user)?

Reply:
Yes, that's what I meant.

Reply:
So would something like the following inside `_check_compute_flag` do?
```
    ew = weight.energy_weight
    if calculator.use_energy and ew < 1e-12:
         logger.warning(msg.format("energy", ew))
```

Reply:
@mjwen Is there a reason we want to keep the argument `residual_data` in `_check_compute_flag`? Because it seems like this function only checks the weights, and now `resudual_data` doesn't have information about the weights anymore.

Reply:
Good catch. You are right. I think we can remove `residual_data` from it now. 

Also, since we are using the new way to assign weight, the [doc](https://kliff.readthedocs.io/en/latest/modules/loss.html#doc-loss) is outdated as well. Can you update the doc as well? [Here](https://github.com/openkim/kliff/blob/master/docs/source/modules/loss.rst) is the source.

Reply:
Sure, I can do that.

So, the `residual_data` pretty much will act like keyword arguments for the residual function then?

Since the primary purpose of `_check_compute_flag` is to check if the weights are large enough and give feedback to user about this, wouldn't it more appropriate to do this check when user creates the weight? That is, what if we move `_check_compute_flag` to weight class and call it inside `weight.compute_weight`?
I also think that this will be more future proof, for example if user wants to define some calculator that doesn't use energy, forces, or stress. In that way, they can define their own `_check_compute_flag` in their weight class.

Reply:
Yes, it is, residual_data is some additional args to residual_fn. Now that only normalize_by_num_atoms is left, I think we can redefine the signature the residual_fn as `residual_fn(prediction, reference, identifier,  weight, natoms, normalize_by_atoms)` to get rid of the use of `residual_data` in all places. Note, this may require the update of some tests and docs. 

> what if we move _check_compute_flag to weight class and call it inside weight.compute_weight?

This may not be easily doable, since it depends of the flags `compute_energy`, `compute_forces`, and `compute_stress`, which are defined in a `Calculator`. However, a `Configuration` is supposed to know nothing of the calculator.

Reply:
> Yes, it is, residual_data is some additional args to residual_fn. Now that only normalize_by_num_atoms is left, I think we can redefine the signature the residual_fn as `residual_fn(prediction, reference, identifier, weight, natoms, normalize_by_atoms)` to get rid of the use of `residual_data` in all places. Note, this may require the update of some tests and docs.

I would still prefer to keep `residual_data`-like argument to be more flexible for the users, for example if users need to use some other information when they define their residual function. Maybe have the signature of the residual_fn as `residual_fn(prediction, reference, identifier, weight, natoms, **residual_data)`.

> This may not be easily doable, since it depends of the flags `compute_energy`, `compute_forces`, and `compute_stress`, which are defined in a `Calculator`. However, a `Configuration` is supposed to know nothing of the calculator.

For this, I was thinking something like the following as a check,
```
if config._forces is not None and np.all(fw < 1e-12):
    logger.warning(msg.format("forces", fw))
```
and suggest to set `use_energy=False` in `calc.create()`, even before users call `calc.create()`.
I have addressed most of your other comments and I can push my commit with an implementation on this weight check.

Reply:
> I would still prefer to keep residual_data-like argument to be more flexible for the users, for example if users need to > use some other information when they define their residual function. Maybe have the signature of the residual_fn as > residual_fn(prediction, reference, identifier, weight, natoms, **residual_data).

The default residual functions we provide can only deal with weights (now moved to the `weight` parameters) and whether to normalize by atoms. If a user wants to do more, a user defined residual_fn has to be provided and the user can do whatever needed. So, it seems not necessary to keep `residual_data`. But I agree that it seems better to keep `residual_data` to give a more controlled way of providing additional info. So, let's do what you've suggested. 

> ```
> if config._forces is not None and np.all(fw < 1e-12):
>      logger.warning(msg.format("forces", fw))
> ```
 
This may lead to some confusion, e.g. in the case where `calculator.use_forces=False`, then the value of `fw` does not matter at all, and this warning is not expected to be shown. But since our default fw is 1.0, hopefully a user playing with `calculator.use_forces` and `fw` will know what they are. I agree with that moving the check to `Configuration` is much cleaner. So, please go ahead and do it.    

> I have addressed most of your other comments and I can push my commit with an implementation on this weight check.

Great!